### PR TITLE
Adapt to the latest elf-edit API changes

### PR DESCRIPTION
**Pull request by travitch**
The ELF Layout type was removed, which the ELF rewriter in renovate used.  The
replacement is explicit re-rendering of the current ELF into a
bytestring (followed by re-parsing the headers).  This change required some
pre-indexing of section and segment offsets and sizes.

Client code is slightly simplified due to the changed API (there is no longer
duplication between 32 and 64 bit architectures).

As part of this change, a logger is explicitly threaded through the rewriter
and analysis.

### CODE REVIEW

**Comment by travitch** [comment]
Actually, it doesn't work yet, but almost does...

File: renovate/utils/run-discovery.hs [hunk 1]
```diff
@@ -24,22 +27,24 @@ main :: IO ()
...
+            EM_PPC64 -> do
+              bin <- loadBinary @PPC defaultLoadOptions e
+              entries <- toList <$> entryPoints bin
+              let pli = ppc64_linux_info bin
+              showDiscoveryInfo $ cfgFromAddrs pli (memoryImage bin) M.empty entries []
+            EM_X86_64 -> do
+              bin <- loadBinary @X86_64 defaultLoadOptions e
+              entries <- toList <$> entryPoints bin
+              showDiscoveryInfo (cfgFromAddrs x86_64_linux_info (memoryImage bin) M.empty entries [])
+            _ -> fail "only X86 and PPC64 supported for now"
```
Comment (kquick) [inline renovate/utils/run-discovery.hs:46]: Good place to use `panic`

Comment (kquick) [reply renovate/utils/run-discovery.hs:46]: There are a couple of `die` calls above.   Is `die` different than `fail` here? Should `die` call `panic`?

File: renovate/src/Renovate/BinaryFormat/ELF.hs [hunk 2]
```diff
@@ -728,22 +791,22 @@ buildNewSymbolTable :: (w ~ MM.ArchAddrWidth arch, E.ElfWidthConstraints w, MM.M
...
+                    -> ElfRewriter lm arch (E.Symtab w)
+buildNewSymbolTable textSecIdx extraTextSecIdx layoutAddr newSyms addrMap baseTable = do
+  let newEnts = newEntries (toMap (E.symtabEntries baseTable))
   let redirections = redirs
-  let tableEntries = V.concat [ E.elfSymbolTableEntries baseTable, V.fromList redirections, newEnts ]
-  return $ baseTable { E.elfSymbolTableEntries = tableEntries
-                     , E.elfSymbolTableIndex = nextSectionIndex elf
+  let tableEntries = V.concat [ E.symtabEntries baseTable, V.fromList redirections, newEnts ]
+  return $ baseTable { E.symtabEntries = tableEntries
+                     -- FIXME: What is this? It seems like it shouldn't be the same as the table size
```
Comment (kquick) [inline renovate/src/Renovate/BinaryFormat/ELF.hs:802]: Should this be fixed now?  If not, should an issue be created for this?

File: renovate/utils/run-discovery.hs (hunk 1 above)
Comment (travitch) [reply renovate/utils/run-discovery.hs:46]: I'll get them all unified eventually - I was trying to not go too off the rails on this PR

File: renovate/src/Renovate/BinaryFormat/ELF.hs (hunk 2 above)
Comment (travitch) [reply renovate/src/Renovate/BinaryFormat/ELF.hs:802]: That is one of the last two questions I need to ask Joe

**Review by kquick — APPROVED** [review]
Other than the comments (and the currently failing CI), this looks good to me.

**Pull request merged**